writer.go

```diff
@@ -231,7 +240,11 @@ func (w *Writer) WriteMessages(ctx context.Context, msgs ...Message) error {
 		timer := time.NewTimer(backoff(attempt+1, 100*time.Millisecond, 1*time.Second))
 		select {
 		case <-timer.C:
-			err = nil
+			// Only clear the error (so we retry the loop) if we have more retries, otherwise
+			// we risk silencing the error.
+			if attempt != w.config.MaxAttempts-1 {
```

Comment:
How about replacing this condition by `attemp < w.config.MaxAttempts`? It is easier to read and more robust (if for any reason attempt happened to be greater than MaxAttempts the current code would break).